src/sentry/eventstream/abstract.py

```diff
@@ -0,0 +1,47 @@
+from abc import ABC, abstractmethod
+from typing import Any, Literal, Mapping, Optional, Tuple, TypeVar, Union
+
+TData = TypeVar("TData")
+
+
+class EventStreamAPI(ABC):
+    @abstractmethod
+    def run_post_process_forwarder(
+        self,
+        consumer_group: str,
+        commit_log_topic: str,
+        synchronize_commit_group: str,
+        commit_batch_size: int,
+        commit_batch_timeout_ms: int,
+        initial_offset_reset: Union[Literal["latest"], Literal["earliest"]],
+    ) -> bool:
+        raise NotImplementedError
+
+    @abstractmethod
+    def requires_post_process_forwarder(self) -> bool:
+        raise NotImplementedError
+
+    @abstractmethod
+    def insert(self, data: TData) -> None:
```

Comment:
The need for a generic here makes me think that either this abstract class should not exist or trying to make it rely on a type that is specific to the implementation is not the right choice.
Generics are meant to make a class parametric, which means make the class work regardless on what the actual type would be, but by preserving type safety. Think about collections. A list is generic to the type of the element so that the same implementation of the List class would work for integers/strings etc. The class never makes a distinction.

In this case the abstract class has no implementation and the subclasses work on one type only and their implementation is type specific. Which means you cannot transparently replace a reference to the abstract class with a reference to any of the children (as they take different and incompatible types), which is what abstract classes are for. This principle is called Liskov Substitution Principle.

So the instrument you are looking for here is not generics. 
I think you have two possible ways:
- remove the abstract class entirely. If the children cannot be replaced with each other it means you may not need that abstraction after all. The abstraction though may be useful for the other methods like run_post_process_forwarder
- Remove the generic and build an abstraction on top of the two message types. Of course this have to be thought through. Your first reaction will likely be that the logic of the two implementations depends on the specifics of the types of the `data` argument. At that point you will have two options: figure out whether you can move the logic that depends on those specifics in `TData` itself or find a common representation of `TData` that allows each implementation to find the data they need. More details on this in the comments below.